cosalib/meta.py: support sub schemas

darkmuggle opened the pull request:
This adds support for discerning whether or not a sub-schema should
exist based on the name in 'meta.json' to support variants.

The new behavior to treat the main schema file as a basename. Any schema
that is found in the same directory with the basename is loaded. The
"name" field is then used to find any sub schemas. For example, given
"v1.json" with a build name of "fedora-coreos", a schema named
"fedora-coreos-v1.json" in the same directory will be applied.

fixtures/fcos.json: defined test JSON for Fedora CoreOS

src/schema/{fedora-coreos,rhcos}-v1.json: stub out for the different
  schemas. This is to support the FCOS use of:
        "fedora-coreos.parent-version" and
        "fedora-coreos.parent-commit" keys.

tests/test_cosalib_meta.py: support testing both FCOS and RHCOS schemas
- looks for all schemas in fixtures directory
- validated the required items against both.

Makefile: set the directory for the test JSON.

darkmuggle commented:
In #1062 @jlebon asked for two "fedora-coreos" keys.  This addresses that need while giving both Fedora CoreOS and RHCOS the ability to define "schema" extensions.

jlebon commented on src/cosalib/meta.py:
```diff
@@ -56,9 +74,10 @@ def validate(self):
+        for k, v in self._validator.items():
+            if k in ("main", self.get("name")):
+                log.debug(f"validating against {k} schema")
+                v.validate(dict(self))
```
Hmm, so this is just doing a sequential check of the various schemas, right? But I think one other really valuable part of validating against a schema is erroring/warning against unknown keys. Otherwise, e.g. it would not have prevented https://github.com/coreos/coreos-assembler/issues/1045 really.

Briefly looking through https://python-jsonschema.readthedocs.io/en/stable/, I couldn't find an option to specify multiple schemas to "overlay", which I think is understandable.

How about something like this:
- we read in the OS-specific schema file, and add the `properties` items to the `properties` of the base schema
- we turn on `additionalProperties: false` in the base schema

darkmuggle replied:
Take another look :)

darkmuggle commented:
I need to fix the tests, but the latest push:
- uses a strict schema (no more extra's allowed)
- uses an enum to determine what is required.

darkmuggle commented:
@jlebon can I bother you for another review? I _think_ this accomplishes what you suggested, while doing strict schema enforcement.

darkmuggle commented:
> I need to fix the tests, but the latest push:
> 
> * uses a strict schema (no more extra's allowed)
> * uses an enum to determine what is required.

Fixed up.  The Schema is now:
1. only allows "fedora-coreos" or "rhcos" as the name
2. based on the name, it uses "anyOf" to enumerate the required elements.

I dropped the sub-schemas and the code around ingesting them

arithx commented on src/schema/v1.json:
```diff
@@ -3,34 +3,83 @@
+           "fedora-coreos.parent-commit",
+           "fedora-coreos.parent-version",
+           "images",
+           "name",
```
Can't `name` be dropped from the `anyOf` of both `rhcos` & `fedora-coreos` because it's in the outer level `required` (line 8)?

cgwalters commented on src/schema/v1.json:
```diff
@@ -872,17 +952,21 @@
+       "fedora-coreos"
       ],
-     "pattern":"^(.*)$"
+     "enum": [
```
It looks like this is exhaustive?  The whole idea is to support people doing custom builds, let's not force `name` to be one of those.

darkmuggle replied:
Changed to support hack builds by defining a "core" schema with the expected outputs for RHCOS and FCOS.

jlebon commented:
Looks like this needs a rebase now.

jlebon commented:
@petr-muller Any idea why the bot is being spammy here?

darkmuggle commented:
@cgwalters @arithx can you gentleman take another look? The schema is conditionalized so that unsupported builds (e.g. non-CoreOS) won't be broken.

arithx commented:
Don't know if it's just github showing the commits wrong but all of the content is in the first commit (`OWNERS: add OWNERS file for PROW integration`)

arithx commented on src/schema/v1.json:
```diff
@@ -3,63 +3,119 @@
+     "examples": [
+       "fedora/x86_64/coreos/testing-devel"
+      ],
+     "pattern":"^(?!\\s*$).+"
```
Optional: single quoting regex patterns to avoid having to escape backslashes can be nice

cgwalters commented on src/schema/v1.json:
```diff
@@ -3,63 +3,121 @@
+    {
+         "if": {
+             "properties": {
+                 "name": { "pattern": "^(rhcos|fedora-coreos)"  }
```
Many of these are core properties emitted by coreos-assembler; I don't see why e.g. if someone is doing a custom build they shouldn't have `coreos-assembler.basearch`.

cgwalters commented on src/schema/v1.json:
```diff
@@ -3,63 +3,121 @@
+                   "coreos-assembler.config-dirty",
+                   "coreos-assembler.config-gitrev",
+                   "coreos-assembler.container-config-git",
+                   "coreos-assembler.container-image-git",
```
Let's not make these required again please?  I don't see a connection between "custom coreos-like builds" and "custom coreos-assembler builds inside an existing development container".

IOW, I want to be able to test patches to coreos-assembler and rpm-ostree inside my dev container, targeting all of fcos/rhcos *and* e.g. [FCOSB](https://github.com/cgwalters/fedora-silverblue-config).

darkmuggle replied:
The problem is that they _are_ required for RHCOS and the purpose of the schema is to validate the results from builds. I am sympathetic towards your pain @cgwalters -- I am -- but your argument is a required production schema should be weakened for you as a developer based on your workflow. 

https://github.com/coreos/coreos-assembler/pull/1089 was targeted at helping you out with this. 

But the larger problem, IMO, is that we have developers who run their own pet container, while COSA expects certain things; that's the tension here. I'm open to ideas

cgwalters replied:
I think delivering coreos-assembler as a container image has been very successful; I often tell people about that.

BUT I don't think we should entirely rule out e.g. shipping it as an RPM too.

And conceptually, these keys only cover the container build path.

And further, they duplicate metadata already available in the container metadata (OpenShift builds e.g. inject the git commit); unfortunately quay.io's build system is more primitive.

So I'm not sure I'd agree these keys are "required for RHCOS".

I guess I can just set the environment variable for now, it'll just be annoying.

jlebon replied:
I'll also chime in as a developer who hacks on cosa from his pet container. :dog2: I think it's important to cover that use-case. Not just for rpm-ostree/`$util`, but just overall I think it yields a faster edit-build-test feedback loop which makes hacking more pleasant. I've documented it [in the README](https://github.com/coreos/coreos-assembler#installing-cosa-inside-an-existing-container) too.

darkmuggle replied:
Dropped in the last push. 

After(if) this merged I'll through up one that allows for per-recipe overrides.

cgwalters commented:
The other aspect to this debate is; maybe we want to drive down into coreos-assembler the concept of an "official" build; we can be more strict in that case.  So it'd be e.g. the pipelines which do `cosa build --official` or something and that adds in the config git schemas rather than triggering off the `name`.

cgwalters commented on src/schema/v1.json:
```diff
@@ -3,63 +3,119 @@
+                   "coreos-assembler.config-gitrev",
+                   "coreos-assembler.image-config-checksum",
+                   "coreos-assembler.image-genver",
+                   "coreos-assembler.image-input-checksum"
```
I think it's OK for all of these to be required too?

Unless we're trying to be compatible with older builds (whole rat's nest there)?

darkmuggle replied:
I was going for the lowest common denominator here. The new schema is strict enough that having these as optional should be good enough IMHO.

darkmuggle commented:
> The other aspect to this debate is; maybe we want to drive down into coreos-assembler the concept of an "official" build; we can be more strict in that case. So it'd be e.g. the pipelines which do `cosa build --official` or something and that adds in the config git schemas rather than triggering off the `name`.

I'll look into it.
